File: src/compiler/core/lexer.ts
// Tipos de tokens da linguagem Zyra
export enum TokenType {
    // Palavras-chave
    IMPORT = 'IMPORT',
    AS = 'AS',
    FROM = 'FROM',
    APP = 'APP',
    SERVER = 'SERVER',
    COMPONENT = 'COMPONENT',
    PAGE = 'PAGE',
    STATE = 'STATE',
    FUNCTION = 'FUNCTION',
    RENDER = 'RENDER',
    IF = 'IF',
    ELSE = 'ELSE',
    FOR = 'FOR',
    IN = 'IN',
    SOCKET = 'SOCKET',
    EMIT = 'EMIT',
    ON = 'ON',
    TRUE = 'TRUE',
    FALSE = 'FALSE',
    RETURN = 'RETURN',
    
    // Símbolos
    LEFT_BRACE = '{',
    RIGHT_BRACE = '}',
    LEFT_PAREN = '(',
    RIGHT_PAREN = ')',
    LEFT_BRACKET = '[',
    RIGHT_BRACKET = ']',
    COLON = ':',
    SEMICOLON = ';',
    COMMA = ',',
    ARROW = '=>',
    DOT = '.',
    
    // Outros
    COMMENT = 'COMMENT',
    STRING = 'STRING',
    NUMBER = 'NUMBER',
    BOOLEAN = 'BOOLEAN',
    IDENTIFIER = 'IDENTIFIER',
    EOF = 'EOF'
}

export interface Token {
    type: TokenType;
    lexeme: string;
    literal: any;
    line: number;
}

export class Lexer {
    private source: string;
    private tokens: Token[] = [];
    private start = 0;
    private current = 0;
    private line = 1;
    
    private keywords: { [key: string]: TokenType } = {
        'import': TokenType.IMPORT,
        'as': TokenType.AS,
        'from': TokenType.FROM,
        'app': TokenType.APP,
        'server': TokenType.SERVER,
        'component': TokenType.COMPONENT,
        'page': TokenType.PAGE,
        'state': TokenType.STATE,
        'function': TokenType.FUNCTION,
        'render': TokenType.RENDER,
        'if': TokenType.IF,
        'else': TokenType.ELSE,
        'for': TokenType.FOR,
        'in': TokenType.IN,
        'socket': TokenType.SOCKET,
        'emit': TokenType.EMIT,
        'on': TokenType.ON,
        'true': TokenType.TRUE,
        'false': TokenType.FALSE,
        'return': TokenType.RETURN
    };

    constructor(source: string) {
        this.source = source;
    }

    scanTokens(): Token[] {
        while (!this.isAtEnd()) {
            this.start = this.current;
            this.scanToken();
        }

        this.tokens.push({
            type: TokenType.EOF,
            lexeme: "",
            literal: null,
            line: this.line
        });

        return this.tokens;
    }

    private scanToken() {
        const c = this.advance();
        
        switch (c) {
            case '(': this.addToken(TokenType.LEFT_PAREN); break;
            case ')': this.addToken(TokenType.RIGHT_PAREN); break;
            case '{': this.addToken(TokenType.LEFT_BRACE); break;
            case '}': this.addToken(TokenType.RIGHT_BRACE); break;
            case '[': this.addToken(TokenType.LEFT_BRACKET); break;
            case ']': this.addToken(TokenType.RIGHT_BRACKET); break;
            case ':': this.addToken(TokenType.COLON); break;
            case ';': this.addToken(TokenType.SEMICOLON); break;
            case ',': this.addToken(TokenType.COMMA); break;
            case '.': this.addToken(TokenType.DOT); break;
            case '=': 
                if (this.match('>')) {
                    this.addToken(TokenType.ARROW);
                }
                break;
            case '/':
                if (this.match('/')) {
                    // Comentário de linha única
                    let comment = '';
                    while (this.peek() != '\n' && !this.isAtEnd()) {
                        comment += this.advance();
                    }
                    this.addToken(TokenType.COMMENT, comment.trim());
                } else if (this.match('*')) {
                    // Comentário multilinhas
                    let comment = '';
                    while (!this.isAtEnd()) {
                        if (this.peek() == '*' && this.peekNext() == '/') {
                            this.advance(); // Consome *
                            this.advance(); // Consome /
                            break;
                        }
                        if (this.peek() == '\n') {
                            this.line++;
                            comment += '\n';
                        } else {
                            comment += this.advance();
                        }
                    }
                    this.addToken(TokenType.COMMENT, comment.trim());
                } else {
                    throw this.error(`Caractere inesperado '${c}'`);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                // Ignora espaços em branco
                break;
            case '\n':
                this.line++;
                break;
            case '"': this.string(); break;
            default:
                if (this.isDigit(c)) {
                    this.number();
                } else if (this.isAlpha(c)) {
                    this.identifier();
                } else {
                    throw this.error(`Caractere inesperado '${c}'`);
                }
                break;
        }
    }

    private string() {
        while (this.peek() !== '"' && !this.isAtEnd()) {
            if (this.peek() === '\n') this.line++;
            this.advance();
        }

        if (this.isAtEnd()) {
            throw new Error(`String não terminada na linha ${this.line}`);
        }

        this.advance();

        const value = this.source.substring(this.start + 1, this.current - 1);
        this.addToken(TokenType.STRING, value);
    }

    private number() {
        while (this.isDigit(this.peek())) this.advance();

        if (this.peek() === '.' && this.isDigit(this.peekNext())) {
            this.advance();
            while (this.isDigit(this.peek())) this.advance();
        }

        this.addToken(
            TokenType.NUMBER,
            parseFloat(this.source.substring(this.start, this.current))
        );
    }

    private identifier() {
        while (this.isAlphaNumeric(this.peek())) this.advance();

        const text = this.source.substring(this.start, this.current);
        const type = this.keywords[text] || TokenType.IDENTIFIER;
        this.addToken(type);
    }

    private isAtEnd(): boolean {
        return this.current >= this.source.length;
    }

    private advance(): string {
        return this.source.charAt(this.current++);
    }

    private addToken(type: TokenType, literal: any = null) {
        const text = this.source.substring(this.start, this.current);
        this.tokens.push({
            type,
            lexeme: text,
            literal,
            line: this.line
        });
    }

    private match(expected: string): boolean {
        if (this.isAtEnd()) return false;
        if (this.source.charAt(this.current) !== expected) return false;

        this.current++;
        return true;
    }

    private peek(): string {
        if (this.isAtEnd()) return '\0';
        return this.source.charAt(this.current);
    }

    private peekNext(): string {
        if (this.current + 1 >= this.source.length) return '\0';
        return this.source.charAt(this.current + 1);
    }

    private isDigit(c: string): boolean {
        return c >= '0' && c <= '9';
    }

    private isAlpha(c: string): boolean {
        return (c >= 'a' && c <= 'z') ||
               (c >= 'A' && c <= 'Z') ||
               c === '_';
    }

    private isAlphaNumeric(c: string): boolean {
        return this.isAlpha(c) || this.isDigit(c);
    }

    private error(message: string): Error {
        return new Error(`${message} na linha ${this.line}`);
    }
} 

File: src/lexer.cpp
#include "lexer.hpp"
#include <stdexcept>
#include <unordered_map>
#include <unordered_set>

namespace zyra {

// Mapa de palavras-chave
static std::unordered_map<std::string, TokenType> keywords = {
    {"component", TokenType::COMPONENT},
    {"state", TokenType::STATE},
    {"style", TokenType::STYLE},
    {"interface", TokenType::INTERFACE},
    {"eventos", TokenType::EVENTOS},
    {"if", TokenType::IF},
    {"else", TokenType::ELSE}
};

// Conjunto de unidades válidas
static std::unordered_set<std::string> units = {
    "%", "px", "rem", "em", "vh", "vw", "s", "ms"
};

Lexer::Lexer(std::string source) : source(std::move(source)) {}

std::vector<Token> Lexer::scanTokens() {
    while (!isAtEnd()) {
        start = current;
        scanToken();
    }
    
    tokens.emplace_back(TokenType::EOF_TOKEN, "", line);
    return tokens;
}

void Lexer::scanToken() {
    char c = advance();
    switch (c) {
        case '{': addToken(TokenType::LEFT_BRACE); break;
        case '}': addToken(TokenType::RIGHT_BRACE); break;
        case '(': addToken(TokenType::LEFT_PAREN); break;
        case ')': addToken(TokenType::RIGHT_PAREN); break;
        case '=': 
            if (match('>')) {
                addToken(TokenType::ARROW);
            } else {
                addToken(TokenType::EQUALS);
            }
            break;
        case '+':
            if (match('=')) {
                addToken(TokenType::PLUS_EQUALS);
            } else {
                throw std::runtime_error("Caractere inesperado '+' na linha " + std::to_string(line));
            }
            break;
        case '-':
            if (match('>')) {
                addToken(TokenType::ARROW);
            } else if (match('=')) {
                addToken(TokenType::MINUS_EQUALS);
            } else if (isDigit(peek())) {
                number();
            } else if (isAlpha(peek()) || peek() == '_') {
                current--; // Volta para incluir o hífen
                identifier();
            } else {
                throw std::runtime_error("Caractere inesperado '-' na linha " + std::to_string(line));
            }
            break;
        case '!': addToken(TokenType::BANG); break;
        case ':': addToken(TokenType::COLON); break;
        case '.': addToken(TokenType::DOT); break;
        case ',': addToken(TokenType::COMMA); break;
        case '#': color(); break;
        case '"': string(); break;
        case '%': addToken(TokenType::UNIT, "%"); break;
        
        // Comentários
        case '/':
            if (match('/')) {
                // Comentário de linha única
                while (peek() != '\n' && !isAtEnd()) advance();
            } else {
                throw std::runtime_error("Caractere inesperado '/' na linha " + std::to_string(line));
            }
            break;
            
        // Ignora espaços em branco
        case ' ':
        case '\r':
        case '\t':
            break;
        case '\n':
            line++;
            break;
            
        default:
            if (isDigit(c)) {
                number();
            } else if (isAlpha(c) || c == '_') {
                identifier();
            } else {
                throw std::runtime_error("Caractere inesperado '" + std::string(1, c) + "' na linha " + std::to_string(line));
            }
            break;
    }
}

void Lexer::string() {
    while (peek() != '"' && !isAtEnd()) {
        if (peek() == '\n') line++;
        advance();
    }
    
    if (isAtEnd()) {
        throw std::runtime_error("String não terminada na linha " + std::to_string(line));
    }
    
    advance(); // Consome as aspas finais
    
    std::string value = source.substr(start + 1, current - start - 2);
    addToken(TokenType::STRING, value);
}

void Lexer::number() {
    while (isDigit(peek())) advance();
    
    if (peek() == '.' && isDigit(peekNext())) {
        advance();
        while (isDigit(peek())) advance();
    }
    
    // Verifica se há uma unidade após o número
    if (peek() == '%') {
        std::string num = source.substr(start, current - start);
        advance(); // Consome o %
        addToken(TokenType::UNIT, num + "%");
        return;
    }
    
    if (isAlpha(peek())) {
        std::string num = source.substr(start, current - start);
        std::string unit;
        while (isAlpha(peek())) {
            unit += advance();
        }
        if (units.find(unit) != units.end()) {
            addToken(TokenType::UNIT, num + unit);
            return;
        }
        // Se não for uma unidade válida, volta atrás
        current -= unit.length();
    }
    
    addToken(TokenType::NUMBER, source.substr(start, current - start));
}

void Lexer::identifier() {
    while (isAlphaNumeric(peek()) || peek() == '-' || peek() == '_') advance();
    
    std::string text = source.substr(start, current - start);
    auto it = keywords.find(text);
    TokenType type = it != keywords.end() ? it->second : TokenType::IDENTIFIER;
    
    addToken(type);
}

void Lexer::color() {
    while (isDigit(peek()) || (peek() >= 'a' && peek() <= 'f') || (peek() >= 'A' && peek() <= 'F')) {
        advance();
    }
    
    addToken(TokenType::COLOR);
}

bool Lexer::isAtEnd() {
    return current >= source.length();
}

char Lexer::advance() {
    return source[current++];
}

bool Lexer::match(char expected) {
    if (isAtEnd() || source[current] != expected) return false;
    current++;
    return true;
}

char Lexer::peek() {
    if (isAtEnd()) return '\0';
    return source[current];
}

char Lexer::peekNext() {
    if (current + 1 >= source.length()) return '\0';
    return source[current + 1];
}

void Lexer::addToken(TokenType type) {
    addToken(type, source.substr(start, current - start));
}

void Lexer::addToken(TokenType type, std::string lexeme) {
    tokens.emplace_back(type, std::move(lexeme), line);
}

bool Lexer::isDigit(char c) {
    return c >= '0' && c <= '9';
}

bool Lexer::isAlpha(char c) {
    return (c >= 'a' && c <= 'z') ||
           (c >= 'A' && c <= 'Z') ||
           c == '_';
}

bool Lexer::isAlphaNumeric(char c) {
    return isAlpha(c) || isDigit(c);
}

} // namespace zyra 